File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        res = float("-inf")

        def dfs(node):
            nonlocal res
            if not node: return 0

            left = dfs(node.left)
            right = dfs(node.right)

            totalSum = sum([node.val, left, right]) 
            leftSum = sum([node.val, left])
            rightSum = sum([node.val, right])
            maxSum = max(node.val, totalSum, leftSum, rightSum)
            if maxSum > res:
                res = maxSum

            return max(node.val + max(left, right), node.val)
    
        dfs(root)
        return res

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxPathSum = function(root) {
    let res = -Infinity;

    const dfs = (node) => {
        if (!node) return 0;

        const left = dfs(node.left);
        const right = dfs(node.right);

        const totalSum = left + right + node.val;
        const leftSum = left + node.val;
        const rightSum = right + node.val;
       
        res = Math.max(res, node.val, leftSum, rightSum, totalSum);
        return Math.max(node.val, node.val + Math.max(left, right));
    }

    dfs(root);
    return res
};